Docsp 9573 monitoring (#88)
mongodb/docs-java: source/fundamentals/monitoring.txt

==========
Monitoring
==========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

.. What is on this page?

> This guide explains how you can set up and configure **monitoring** in the

Author: I believe this intro style is also used in our documentation. I prefer the intro as written, so I will not implement the suggested edit.

See the following pages for examples of this intro style:

[POJO](https://docs-mongodbcom-staging.corp.mongodb.com/e1acf1b/java/docsworker-xlarge/DOCSP-9573-Monitoring/fundamentals/data-formats/document-data-format-pojo/)
[BSON](https://docs-mongodbcom-staging.corp.mongodb.com/e1acf1b/java/docsworker-xlarge/DOCSP-9573-Monitoring/fundamentals/data-formats/document-data-format-bson/)

MongoDB Java driver.

.. What do any new terms mean?

> Monitoring is the process of getting information about the activities a piece of

Author: I agree that "piece of software" is too wordy. I am changing it to "program".

software performs. You can use the information you get from monitoring in your
application or an application performance management library. 

Monitoring the MongoDB Java driver lets you understand the
driver's resource usage and performance, and can help you make informed
decisions when designing and debugging your application.    

.. What can you expect to see on this page?

> On this page you will learn how to perform these tasks:

Author: I don't think the "guide"/"page" discrepancy is a problem. Our sidebar TOC refers to all of our guides/pages as "pages" (says "On This Page" for as far as I can tell all of our pages), but we often refer to them as guides in the intro. A "guide" is necessarily a "page" within our docs.

I do not agree with the second edit. I believe the proposed rephrasing is less task oriented than the current phrasing on the page (I believe readers are interested in "tasks" they wish to perform rather than "sections" they may read).

> On this page you will learn how to perform these tasks:

Author: This edit suggests it may be beneficial to discuss how we would like to refer to our pages. Is "guide" the only appropriate term? Are "page" and "guide" both acceptable?

- :ref:`Monitor different types of events in the MongoDB Java Driver <monitoring-monitor-events>`
- :ref:`Monitor connection pool events with Java Management Extensions (JMX) and JConsole <monitoring-jmx>`

This page shows how to use information about the activity of the driver in code.
If you would like to learn how to record events the driver performs,
consider reading our :doc:`guide on logging </fundamentals/logging>`.

.. _monitoring-monitor-events:

Monitor Events
--------------

To monitor an **event**, you must register a **listener** on your ``MongoClient``
instance. 

> An event is any action that happens in code. The driver includes functionality

Author: 1: By action I mean anything that happens in a program (variable is referenced, network request is made, resource is allocated). An event is anything that happens, so "action" is very broad. 

2: I am referring to the code and not to the database.

> An event is any action that happens in code. The driver includes functionality

Author: I'll change "in code" to "in a running program". This seems I think this is similar to the edit made here https://github.com/mongodb/docs-java/pull/88#discussion_r643139809

for listening to a subset of the events that occur when the driver is running.

A listener is a class that performs some action when certain events occur. All
listener classes extend the ``EventListener`` interface. A listener's
API defines the events it can respond to.

Each method of a listener class represents a response to a certain event. Each
method receives one argument: an object representing the event the method
responds to. All event objects are a subclass of the ``EventObject`` class.  

The MongoDB Java driver organizes the events it defines into three categories:

> - Command Events

Author: I agree that this could be helpful for some readers, but I believe that turning these items into links would[ not provide sufficient context](https://docs.mongodb.com/meta/style-guide/style/links-and-cross-references/#link-and-cross-reference-text).

> Begin a cross-reference sentence with the purpose or benefit of the cross-reference (such as more information or examples). Such context helps users decide whether to follow the reference.

I do not wish to rephrase this sentence to be about the page itself (I want this sentence to be focused on the driver rather than the page), so I would like to keep this as it is currently written (not linking).

> - Command Events

Reviewer: Since it's only 3 items and none are being defined here, can it be in a sentence? This would save vertical space.

> - Command Events

Author: The style guide wants lists of three or more items broken out of paragraphs into lists.

[Section of Style Guide](https://docs.mongodb.com/meta/style-guide/writing/write-clear-concise-sentences-paragraphs/#create-short-paragraphs)

>  When listing three or more items, use a bullet list instead of embedding the items in a paragraph.

I also personally prefer this style, as I think it is easier on the eyes.

Based off of this section of the style guide, I think it is correct as written.

- Cluster Events
- Connection Pool Events

The following sections show how to monitor each event category.

For a full list of the events you can monitor,
:java-docs:`see the event package of the MongoDB Java Driver <apidocs/mongodb-driver-core/com/mongodb/event/package-summary.html>`.

For more information on the classes and interfaces mentioned in this section,
see the following API documentation:

- `EventListener <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventListener.html>`__
- `EventObject <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventObject.html>`__

Command Events
~~~~~~~~~~~~~~

A command event is an event related to a MongoDB database command. Here are some
examples of database commands that produce command events:

- ``find``
- ``insert``
- ``delete``

> - ``count`` 

Author: I think it is correct as written. See the [commands reference page](https://docs.mongodb.com/manual/reference/command/) (commands are written in same format as here). 

From the Glossary:

[Database Command](https://docs.mongodb.com/manual/reference/glossary/#std-term-database-command): A MongoDB operation, other than an insert, update, remove, or query. For a list of database commands, see Database Commands. To use database commands, see Issue Commands.

[Operator](https://docs.mongodb.com/manual/reference/glossary/#std-term-operator): A keyword beginning with a $ used to express an update, complex query, or data transformation. For example, $gt is the query language's "greater than" operator. For available operators, see Operators.

Based off of these definitions, operators (keywords used in update or query specifications) receive a `$`, and methods receive `()`.

   :language: java
   :dedent:
   :start-after: start monitor-command-example
   :end-before: end monitor-command-example

The above code snippet should produce output that looks like this:

.. code-block:: none
   :copyable: false

   Command 'find' took 19 milliseconds
   Command 'aggregate' took 15 milliseconds
   Command 'endSessions' took 14 milliseconds

For more information on the classes and methods mentioned in this section, see
the following API documentation:

- :java-docs:`CommandListener <apidocs/mongodb-driver-core/com/mongodb/event/CommandListener.html>`
- :java-docs:`MongoClientSettings <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html>`
- :java-docs:`MongoClient <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html>`
- :java-docs:`CommandStartedEvent <apidocs/mongodb-driver-core/com/mongodb/event/CommandStartedEvent.html>`
- :java-docs:`CommandSucceededEvent <apidocs/mongodb-driver-core/com/mongodb/event/CommandSucceededEvent.html>`
- :java-docs:`CommandFailedEvent <apidocs/mongodb-driver-core/com/mongodb/event/CommandFailedEvent.html>`

Cluster Events
~~~~~~~~~~~~~~

> A cluster event in the driver corresponds to a change in the state of the

Author: I believe this comment responds to this issue: https://github.com/mongodb/docs-java/pull/88#discussion_r643284081

MongoDB instance or cluster you have connected the driver to. Cluster events are

> also called server discovery and monitoring (SDAM) events. 

Author: I disagree with this edit. I don't believe it would benefit readers to see the names as equivalent (cluster name is more useful in the Java driver, so I believe it should be given preference). The second name is introduced later only because the it is the term used in the specification, and I don't want readers to be confused when they are linked to the specification later in the paragraph.

> The driver defines nine cluster events. The driver divides these nine events
> between three separate listener interfaces which each listen for three of the
> nine events:

Reviewer: ```suggestion
The driver defines nine cluster events and divides them into three listener interfaces. Each interface listens for three of the nine events:
```

Author: I am not sure how this edit improves the piece. This seems to duplicate "interface" rather than "the driver" in breaking this across multiple sentences. 

Am I missing the crux of this edit in some way?

> - ``ClusterListener``: :spec:`topology </server-discovery-and-monitoring/server-discovery-and-monitoring.rst#topology>`

Reviewer: Suggestion: As is, it seems somewhat abrupt. I didn't realize you were describing what each listener listens for.
Applies to all instances.
```suggestion
- ``ClusterListener``: listens for :spec:`topology </server-discovery-and-monitoring/server-discovery-and-monitoring.rst#topology>`
```

Author: Thanks for the note. I agree this transition is abrupt. Fix:

Going forward, I will work on smoothing my transitions. I believe this is becoming a theme in reviews (choppy sentences and abrupt transitions).

  related events
- ``ServerListener``: events related to ``mongod`` or ``mongos`` processes
- ``ServerMonitorListener``: heartbeat related events

For a detailed description of all nine cluster events in the driver, see the 
:spec:`MongoDB SDAM monitoring specification </server-discovery-and-monitoring/server-discovery-and-monitoring-monitoring.rst#events>`.

> 

Reviewer: Issue: Since you include a sentence about what the user needs to do to create that event in the other two event sections, I think you should include one here as well.

Author: Thanks for the note. I agree the section could benefit from addition. Fix: https://github.com/mongodb/docs-java/pull/88/commits/cda95419a82eac050d8c18b114c719d79df64b43

Going forward, I will try and better sense parallel structure issues.

``IsReadAndWrite`` class.

.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
   :language: java
   :dedent:
   :start-after: start monitor-cluster-example
   :end-before: end monitor-cluster-example

The above code snippet should produce output that looks like this:

.. code-block:: none
   :copyable: false

   Able to write to server
   Able to read from server

For more information on the classes and methods mentioned in this section, see
the following API documentation:

- :java-docs:`ClusterListener <apidocs/mongodb-driver-core/com/mongodb/event/ClusterListener.html>`
- :java-docs:`ServerListener <apidocs/mongodb-driver-core/com/mongodb/event/ServerListener.html>`
- :java-docs:`ServerMonitorListener <apidocs/mongodb-driver-core/com/mongodb/event/ServerMonitorListener.html>`
- :java-docs:`MongoClientSettings <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html>`
- :java-docs:`MongoClient <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html>`
- :java-docs:`ClusterDescriptionChangedEvent <apidocs/mongodb-driver-core/com/mongodb/event/ClusterDescriptionChangedEvent.html>`

Connection Pool Events
~~~~~~~~~~~~~~~~~~~~~~

> Connection pool events are events related to a **connection pool** held by the driver.

Author: Thanks for the note. I have updated all event subsections to have a parallel intro sentence structure. Fix: https://github.com/mongodb/docs-java/pull/88/commits/3e00609af4b22c26daa2affc1596ecdf416f3536

Going forward, I will continue to try and refine my ear for parallel structure between sections.

To monitor connection pool events, write a class that implements the
``ConnectionPoolListener`` interface and register an instance of that class with your
``MongoClient`` instance.

Example
^^^^^^^

This example shows how to make a listener class that prints a message each time
you check out a connection from your connection pool.

The following code defines the ``ConnectionPoolLibrarian`` class which implements the
``ConnectionPoolListener`` interface.

.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
   :language: java
   :dedent:
   :start-after: start cp-listener-impl
   :end-before: end cp-listener-impl

The following code adds an instance of the ``ConnectionPoolLibrarian`` class to a
``MongoClient`` object. The code then runs a database command to test the
librarian.

.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
   :language: java
   :dedent:
   :start-after: start monitor-cp-example
   :end-before: end monitor-cp-example

The above code snippet should produce output that looks like this:

.. code-block:: none
   :copyable: false

   Let me get you the connection with id 21...
   Let me get you the connection with id 21...

> It looks like our find operation requested two connections from the pool.
> If you are curious why the driver uses two connections in one find operation, 
> `you can find the answer in this line of the MongoDB driver source code <https://github.com/mongodb/mongo-java-driver/blob/a82ce71d0aa7da9c6455f1b52ae24ed3cd4c1f62/driver-sync/src/main/com/mongodb/client/internal/FindIterableImpl.java#L203>`__.

Reviewer: ```suggestion 
Our find operation requested two connections from the pool.
For information on why the driver uses two connections in a find operation, see
`this line of the MongoDB driver source code <https://github.com/mongodb/mongo-java-driver/blob/a82ce71d0aa7da9c6455f1b52ae24ed3cd4c1f62/driver-sync/src/main/com/mongodb/client/internal/FindIterableImpl.java#L203>`__.
```

Author: I agree that "our find operation" is incorrect. I disagree with the other two edits. This is meant to be a fun aside for readers exploring a curious behavior of the driver, not necessary for understanding the feature being described. For that reason, this was written with a different tone, "it looks like... if you are curious".

```suggestion
It looks like the above find operation requested two connections from the pool.
If you are curious why the driver uses two connections in one find operation, 
`you can find the answer in this line of the MongoDB driver source code <https://github.com/mongodb/mongo-java-driver/blob/a82ce71d0aa7da9c6455f1b52ae24ed3cd4c1f62/driver-sync/src/main/com/mongodb/client/internal/FindIterableImpl.java#L203>`__.
```

For more information on the classes and methods mentioned in this section, see
the following API documentation:

- :java-docs:`ConnectionPoolListener <apidocs/mongodb-driver-core/com/mongodb/event/ConnectionPoolListener.html>`
- :java-docs:`MongoClientSettings <apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html>`
- :java-docs:`MongoClient <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html>`
- :java-docs:`ConnectionCheckedOutEvent <apidocs/mongodb-driver-core/com/mongodb/event/ConnectionCheckedOutEvent.html>`
- :java-docs:`ConnectionCheckOutFailedEvent <apidocs/mongodb-driver-core/com/mongodb/event/ConnectionCheckOutFailedEvent.html>`

.. _monitoring-jmx:

> Monitor Connection Pools with JMX

Reviewer: Question: Should this heading say: "Monitor Connection Pool Events with JMX"?

Author: Thanks for the note. I believe it should, good catch.

---------------------------------

You can monitor connection pool events using **Java Management Extensions (JMX)**. 

> JMX provides tools to monitor applications and devices.

Reviewer: Issue: This sentence leaves me wondering what tools? And in what type of situation is it useful?

Author: I feel it is appropriate to provide an unsatisfactory definition of JMX here as it is not the intention of this guide to explain JMX to our readers. The goal of this definition is to give readers enough context to consult Oracle documentation on JMX.

For more information on JMX, see 
`the official Oracle JMX documentation <https://docs.oracle.com/javase/tutorial/jmx/index.html>`__. 

> JMX Support in the MongoDB Driver

Reviewer: Suggestion: I think the MongoDB Driver is implied here.
```suggestion
JMX Support
```

Author: Thanks for the note. I agree that this is unnecessary. Thank you for pointing this out.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To enable JMX connection pool monitoring, add an instance of the
``JMXConnectionPoolListener`` class to your ``MongoClient``.

The ``JMXConnectionPoolListener`` class performs the following actions: 

> - Creates MXBean instances for each ``mongod`` or ``mongos`` process the driver

Reviewer: Question: Should `MXBean` and `MBean` be monospaced? 
Applies to all instances.

Author: I do not believe they should. These are referring to a software design concept rather than a class. I believe they should receive the same formatting as POJO (no monospace).

  maintains a connection pool with.

> - Registers these MXBean instances with the platform MBean server. 

Reviewer: ```suggestion
- Registers the MXBean instances with the platform MBean server. 
```

Author: I believe "these" is correct here. 

Definition of [these](https://www.google.com/search?q=these&oq=these&aqs=chrome..69i57j69i60l5j69i65j69i64.502j0j7&sourceid=chrome&ie=UTF-8): "referring to a specific thing or situation just mentioned."

Is there a reason why "the" is more appropriate here? Am I missing the crux of this suggestion?

Author: Sounds good. 

Reasons for keeping it as is:

I don't think this should be numbered, as in code this all happens as one function call, but I think it should be "these" as it is referencing the items created in the previous bullet.

> MXBeans registered on the platform MBean server have the following properties:

Reviewer: Issue: Before showing the properties of MXBean, I think it would be helpful to briefly explain what it is/does.

.. list-table::
   :header-rows: 1

>    :stub-columns: 1

Reviewer: Issue: the columns have unsatisfactory widths

Author: Thanks for the note, I agree these columns are too narrow. Fix: https://github.com/mongodb/docs-java/pull/88/commits/fa78aebb61993631209080b0147246e5cc78d347

   * - Property

>      - Description

Reviewer: Question: Do any of the properties have default values? If so, I think it's worth listing as the last line in the description column,

Author: I believe these are set through getter/setter methods on the connection pool. They would be default connection pool values, which I do not believe should be documented here.

Author: I think these are different cases.

Kafka is documenting a feature which can be configured.

This is documenting a feature which exposes values to a monitoring framework. This feature in and of itself does not configure connection pools, it only passes its values. I agree that a discussion of connection pools should document default connection pool values, but I think this (a discussion of a feature exposing connection pool values to JMX) should not discuss default connection pool values as it is not immediately tied to the feature.

   * - ``port``
     - The port on which the ``mongod`` or ``mongos`` process is listening.

   * - ``minSize``
     - The minimum size of the connection pool, including idle and in-use connections.

   * - ``maxSize``
     - The maximum size of the connection pool, including idle and in-use connections.

   * - ``size``
     - The current size of the connection pool, including idle and in-use connections.

   * - ``checkedOutCount``
     - The current count of connections that are in use.

All MXBean instances created by the driver are under the domain
``"org.mongodb.driver"``.

For more information on the topics discussed in this subsection, see the
following resources:

- `Oracle Platform MBean Server Reference documentation <https://docs.oracle.com/en/java/javase/16/management/overview-java-se-monitoring-and-management.html#GUID-F7B9AB8A-F5A8-472A-AEC6-93B5B7FBE7CE>`__
- `Oracle MXBean documentation <https://docs.oracle.com/javase/tutorial/jmx/mbeans/mxbeans.html>`__

JMX and JConsole Example
~~~~~~~~~~~~~~~~~~~~~~~~

This example shows how you can monitor the driver's connection pools using JMX
and **JConsole**. JConsole is a JMX compliant GUI monitoring tool that comes with

> the Java Platform.

Reviewer: Question: Is there a reason to include JConsole here? If so, I think there needs to be more information around why it's useful and when to use it before jumping into an example.

Author: I believe some discussion of JConsole is necessary on this page. JMX requires a JMX compliant monitoring tool, the easiest to use JMX compliant tool [(and the tool referenced by Oracle in it's JMX monitoring guide ](https://docs.oracle.com/en/java/javase/16/management/monitoring-and-management-using-jmx-technology.html#GUID-E17343B0-FED2-4050-88D7-E4793E704ED5)) is JConsole. For this reason, I believe JConsole must be mentioned.

Author: I agree I think this makes sense. I have expanded the disclaimer to also include JMX and moved it to the top of the example.

Fix:

The following code snippet adds a ``JMXConnectionPoolListener`` to a
``MongoClient`` instance. The code then starts a JMX connection server exposing
the platform MBean server on port 9999 for JConsole to connect to.

.. literalinclude:: /includes/fundamentals/code-snippets/JMXMonitoring.java
   :language: java
   :dedent:
   :start-after: start jmx-example
   :end-before: end jmx-example

The above code snippet should produce output that looks like this:

.. code-block:: none
   :copyable: false

   Start the connector server...
   The connector server started.
   Press <Enter> to stop the server.

Once you have started your server, open JConsole in your terminal using the
following command:

.. code-block:: shell
   
   jconsole

> .. note:: Consult the Official JConsole Documentation
> 
>    The descriptions of the JConsole GUI on this page are illustrative
>    rather than a source of truth. For guaranteed up to date information, consult 
>    `the official JConsole documentation <https://www.oracle.com/technical-resources/articles/java/jconsole.html>`__. 

Reviewer: Issue: Let's show confidence in our documentation. Instead, you can omit this or put this after your steps (at line 395) to say:

```suggestion
For more information on starting up and using JConsole, visit `the official JConsole documentation <https://www.oracle.com/technical-resources/articles/java/jconsole.html>`__. 
```

Author: As JConsole is a third party tool that may change without us knowing, I believe it is appropriate to provide a disclaimer that readers should expect discrepancies in our guide and their experience.

In a previous review documenting 3rd party technologies, [a disclaimer similar to this was requested by Jeff Yemin](https://github.com/mongodb/docs-java/pull/76#discussion_r623846300)

==========
Monitoring
==========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

.. What is on this page?

This guide explains how you can set up and configure **monitoring** in the
MongoDB Java driver.

.. What do any new terms mean?

Monitoring is the process of getting information about the activities a running 

> program performs for use in an application or an application performance

Reviewer: ```suggestion
program performs to use in an application or an application performance
```

Author: I made a sentence that I think is equivalent to test this edit:

"Dish washing is the process of washing dishes for use in meals."
"Dish washing is the process of washing dishes to use in meals."

I looked around online and I think the main difference is that  "use in meals" becomes a noun with "for" and "use" becomes a verb with "to". 

In my head "for" seems like a slightly better fit, but I think both are correct. Because I can't see a reason to choose either one, I will go with the one I prefer and stick with "for".

management library. 

Monitoring the MongoDB Java driver lets you understand the
driver's resource usage and performance, and can help you make informed
decisions when designing and debugging your application.    

.. What can you expect to see on this page?

In this guide you will learn how to perform these tasks:

- :ref:`Monitor different types of events in the MongoDB Java Driver <monitoring-monitor-events>`
- :ref:`Monitor connection pool events with Java Management Extensions (JMX) and JConsole <monitoring-jmx>`

This guide shows how to use information about the activity of the driver in code.
If you would like to learn how to record events the driver performs,
consider reading our :doc:`guide on logging </fundamentals/logging>`.

.. _monitoring-monitor-events:

Monitor Events
--------------

To monitor an **event**, you must register a **listener** on your ``MongoClient``
instance. 

An event is any action that happens in a running program. The driver includes functionality
for listening to a subset of the events that occur when the driver is running.

A listener is a class that performs some action when certain events occur. All

> listener classes extend the ``EventListener`` interface. A listener's

Reviewer 2: > All listener classes extend the ``EventListener`` interface

Not true in practice. The driver's event listener interfaces are not.

Author: Thanks for the note. Apologies, I didn't realize that this convention was not always followed in practice. I am removing content related to this from the page.

Purely for my own edification (feel free to ignore):

Why are the driver's events and listeners not subclasses of `EventListener` and `EventObject`? Just curious why this decision was made (no need to answer this if you are busy).

Reviewer 2: I think those types go back to an early time, when listeners were used primarily in GUI applications.  They are just not useful in this context.

API defines the events it can respond to.

Each method of a listener class represents a response to a certain event. Each
method receives one argument: an object representing the event the method

> responds to. All event objects are a subclass of the ``EventObject`` class.  

Reviewer 2: > All event objects are a subclass of the ``EventObject`` class

Not true in practice.  The driver's event objects are not.

Author: Response is here: https://github.com/mongodb/docs-java/pull/88#discussion_r646715047

The MongoDB Java driver organizes the events it defines into three categories:

- Command Events
- Server Discovery and Monitoring Events
- Connection Pool Events

The following sections show how to monitor each event category.

For a full list of the events you can monitor,
:java-docs:`see the event package of the MongoDB Java Driver <apidocs/mongodb-driver-core/com/mongodb/event/package-summary.html>`.

For more information on the classes and interfaces mentioned in this section,
see the following API documentation:

- `EventListener <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventListener.html>`__
- `EventObject <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventObject.html>`__

> Command Events

Reviewer 2: Add a note indicating that the driver does not publish events for commands called internally by the driver, e.g. for monitoring the cluster or for handshake on connection establishment, which includes `hello` and auth-related commands.

Author: Thanks for the suggestion. Here is the note: https://github.com/mongodb/docs-java/pull/88/commits/7bd930288fa4d60fbd8e9520f5c414b5602e658e

Let me know if you would like this note to go into more detail. I tried digging into the driver code to get an understanding of the specific types of commands the driver would not produce events for, but had trouble coming up with anything useful to readers. Apologies for not being able to fill out this content on my own, and please let me know if this is insufficient.

Reviewer 2: This looks just fine.  Thanks.

~~~~~~~~~~~~~~

A command event is an event related to a MongoDB database command. Some
examples of database commands that produce command events are ``find``,
``insert``, ``delete``, and ``count``. 

To monitor command events, write a class that implements the
``CommandListener`` interface and register an instance of that class with your
``MongoClient`` instance.

For more information on MongoDB database commands, see the
:manual:`MongoDB manual entry on database commands </reference/command/>`.

.. note:: Redacted Output

>    As a security measure, the driver is unable to monitor some commands. This

Reviewer 2: > unable

Consider a different word.  The driver could monitor all commands.  We just chose not to.  Also, it's not that we don't monitor them, we just redact the command documents.  The events are still generated.

Author: Thanks for the note. I agree "unable" is the incorrect word. I have rephrased this note to change language around "unable" and "can't monitor" to be more focused on redacting contents. Fix: https://github.com/mongodb/docs-java/pull/88/commits/4cfc007205c993ad583f9d72594b84cbde7f7d4f

   protects the sensitive information contained in the output of these
   commands. For a full list of commands the driver can't monitor,
   see the 
   :spec:`MongoDB command monitoring specification </command-monitoring/command-monitoring.rst#security>`.

Example
^^^^^^^

> This example shows how to make a timer for database commands. The timer makes it so

Reviewer 2: Unfortunately this not a good example, because both CommandSucceededEvent and CommandFailedEvent include an elapsedTime property!

Reviewer 2: Also, if this is all you want to do, better to enable DEBUG logging, which already logs elapsed time.

Author: [Facepalm](https://en.wikipedia.org/wiki/Facepalm)

Thanks for pointing this out. I agree that as there are much more straightforward ways to achieve this, this example should be changed. I have changed this example to be a command counter (keeps track of how many times each command has been successfully executed in a map and prints the map). Fix: https://github.com/mongodb/docs-java/pull/88/commits/c0fecb8671ddbe9318a92b1c759220198d4bfb04

Reviewer 2: This is a better example.  Only suggestion is to make the `commandSucceeded` method` `synchronized`, since as is the listener will suffer from race conditions in a multi-threaded application.

that every time the driver executes a command on your MongoDB instance
you see a message letting you know how long it took to execute the command.

To make a timer, do the following:

#. Make a class with timer functionality that implements the ``CommandListener`` interface.
#. Add an instance of the new class that implements ``CommandListener`` to a ``MongoClientSettings`` object.
#. Configure a ``MongoClient`` instance with the ``MongoClientSettings`` object.

> The following code defines the ``CommandTimer`` class which implements the

Reviewer: ```suggestion
The following code defines the ``CommandTimer`` class, which implements the
```
Also applies to the other 2 event examples below.

Author: I believe the suggested edit should not be performed. I think "which implements the ____ interface" is a restrictive clause. I believe you only [add a comma before which for nonrestrictive clauses. ](https://www.grammarly.com/blog/comma-before-which/?&utm_source=google&utm_medium=cpc&utm_campaign=10273012991&utm_targetid=dsa-1233402314764&gclid=Cj0KCQjw2NyFBhDoARIsAMtHtZ7H64TbNm032HTN7BNzgoBZlmVoB8hKBXbACGNcXt6xhoHi3iiAg2saAp43EALw_wcB&gclsrc=aw.ds)

To monitor connection pool events, write a class that implements the
``ConnectionPoolListener`` interface and register an instance of that class with your
``MongoClient`` instance.

Example
^^^^^^^

This example shows how to make a listener class that prints a message each time
you check out a connection from your connection pool.

The following code defines the ``ConnectionPoolLibrarian`` class which implements the
``ConnectionPoolListener`` interface.

.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
   :language: java
   :dedent:
   :start-after: start cp-listener-impl
   :end-before: end cp-listener-impl

The following code adds an instance of the ``ConnectionPoolLibrarian`` class to a
``MongoClient`` object. The code then runs a database command to test the
librarian.

.. literalinclude:: /includes/fundamentals/code-snippets/Monitoring.java
   :language: java
   :dedent:
   :start-after: start monitor-cp-example
   :end-before: end monitor-cp-example

The above code snippet should produce output that looks like this:

.. code-block:: none
   :copyable: false

   Let me get you the connection with id 21...
   Let me get you the connection with id 21...

> It looks like the above find operation requested two connections from the pool.

Reviewer 2: That's not actually the reason.   Here's the stack trace for the second check-out:

```
connectionCheckedOut:159, ConnectionPoolLibrarian (com.mycompany.app)
get:123, DefaultConnectionPool (com.mongodb.internal.connection)
get:100, DefaultConnectionPool (com.mongodb.internal.connection)
getConnection:96, DefaultServer (com.mongodb.internal.connection)
endClosedSessions:156, ServerSessionPool (com.mongodb.internal.session)
close:103, ServerSessionPool (com.mongodb.internal.session)
close:110, MongoClientDelegate (com.mongodb.client.internal)
close:139, MongoClientImpl (com.mongodb.client.internal)
monitorConnectionPoolEvent:81, Monitoring (com.mycompany.app)
main:28, Monitoring (com.mycompany.app)
```

So it's happening because of a requirement that the driver informs the server about sessions that it has created and will no longer be using, since the client is being closed.

You can avoid this by not calling MongoClient.close()

Author: Thanks for catching this. Apologies for getting this incorrect. Here is the fix: https://github.com/mongodb/docs-java/pull/88/commits/c7ed0891dc61a14408a0085f0125838bc058ac9d

Going forward, I will try check these sorts of statements by consulting stack traces before including them in pages. Thanks for sharing this technique.

Reviewer 2: It's ok.  You're learning!

The following code snippet adds a ``JMXConnectionPoolListener`` to a
``MongoClient`` instance. The code then starts a JMX connector server exposing
the platform MBean server on port 9999 for JConsole to connect to.

.. literalinclude:: /includes/fundamentals/code-snippets/JMXMonitoring.java
   :language: java
   :dedent:
   :start-after: start jmx-example
   :end-before: end jmx-example

The above code snippet should produce output that looks like this:

.. code-block:: none
   :copyable: false

   Start the connector server...
   The connector server started.
   Press <Enter> to stop the server.

Once you have started your server, open JConsole in your terminal using the
following command:

.. code-block:: shell
   
   jconsole

Once JConsole is open, perform the following actions in the GUI:

#. Select the Java process running the above example code.
#. Press :guilabel:`Insecure Connection` in the warning dialog box.
#. Click on the :guilabel:`MBeans` tab.
#. Inspect your connection pool events under the ``"org.mongodb.driver"`` domain.

When you no longer want to inspect your connection pools in JConsole, do the
following: 

> - Exit JConsole by closing the JConsole window

Reviewer: Nit: Since you number the other sections for steps, you should here too.

Author: As these actions are not ordered (they can be performed in any order), I do not believe these should be numbered.

==========
Monitoring
==========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

.. What is on this page?

This guide explains how you can set up and configure **monitoring** in the
MongoDB Java driver.

.. What do any new terms mean?

Monitoring is the process of getting information about the activities a
program performs. You can use the information you get from monitoring in your

> application or an application performance management library. 

Author: I don't think this is an incomplete sentence.

Equivalent sentence: "You can use a screwdriver in your house."

- Subject: you
- Verb: use

I don't think this paragraph should touch on how monitoring information can be used (monitoring is mainly about getting information from running code to use in your code). 

I think describing how monitoring you can use monitoring in the driver is appropriate.

Each method of a listener class represents a response to a certain event. Each
method receives one argument: an object representing the event the method
responds to. All event objects are a subclass of the ``EventObject`` class.  

The MongoDB Java driver organizes the events it defines into three categories:

- Command Events
- Cluster Events
- Connection Pool Events

The following sections show how to monitor each event category.

For a full list of the events you can monitor,
:java-docs:`see the event package of the MongoDB Java Driver <apidocs/mongodb-driver-core/com/mongodb/event/package-summary.html>`.

For more information on the classes and interfaces mentioned in this section,
see the following API documentation:

- `EventListener <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventListener.html>`__
- `EventObject <https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/EventObject.html>`__

Command Events
~~~~~~~~~~~~~~

A command event is an event related to a MongoDB database command. Here are some
examples of database commands that produce command events:

- ``find``
- ``insert``
- ``delete``

> - ``count`` 

Author: I believe this comment addresses this suggestion: https://github.com/mongodb/docs-java/pull/88#discussion_r643971348

- :java-docs:`MongoClient <apidocs/mongodb-driver-sync/com/mongodb/client/MongoClient.html>`
- :java-docs:`CommandStartedEvent <apidocs/mongodb-driver-core/com/mongodb/event/CommandStartedEvent.html>`
- :java-docs:`CommandSucceededEvent <apidocs/mongodb-driver-core/com/mongodb/event/CommandSucceededEvent.html>`
- :java-docs:`CommandFailedEvent <apidocs/mongodb-driver-core/com/mongodb/event/CommandFailedEvent.html>`

Cluster Events
~~~~~~~~~~~~~~

A cluster event is an event related to a change in the state of the
MongoDB instance or cluster you have connected the driver to. Cluster events are
also called server discovery and monitoring (SDAM) events. 
 
The driver defines nine cluster events. The driver divides these nine events
between three separate listener interfaces which each listen for three of the
nine events. Here are the three interfaces and the events they listen for:

- ``ClusterListener``: :spec:`topology </server-discovery-and-monitoring/server-discovery-and-monitoring.rst#topology>`
  related events
- ``ServerListener``: events related to ``mongod`` or ``mongos`` processes
- ``ServerMonitorListener``: heartbeat related events

> To monitor one of the three types of cluster events, write a class that

Reviewer: Nit: I don't think mentioning "one of the three" is necessary since you just listed them
```suggestion
To monitor any of the cluster events, write a class that
```

Author: The reason this phrasing was chosen is that a Java class can implement multiple interfaces. The way a reader has to attach an event to their MongoClient it wouldn't make sense to do this, so I wanted to be explicit that you should only implement classes that monitor one type at a time.

implements one of the three interfaces above and register an instance of that
class with your ``MongoClient`` instance.

> For a detailed description of all nine cluster events in the driver, see the 

Reviewer: Nit: Didn't think you needed to mention nine here since you did above.
```suggestion
For a detailed description for each cluster event in the driver, see the 
```

Author: I agree "of all nine" can be removed here.

```suggestion
For a detailed description of each cluster event in the driver, see the 
```
